torchmultimodal/modules/layers/transformer.py

```diff
@@ -0,0 +1,327 @@
+        layer_output = self.output(layer_output)
+        layer_output = self.dropout(layer_output)
+        layer_output += hidden_states
+
+        outputs = (layer_output,) + outputs
+
+        return outputs
+
+
+class TransformerEncoder(nn.Module):
+    def __init__(
+        self,
+        hidden_size: int = 768,
+        num_attention_heads: int = 12,
+        num_hidden_layers: int = 12,
+        hidden_dropout_prob: float = 0.0,
+        intermediate_size: int = 3072,
+        intermediate_activation: Callable[..., Tensor] = nn.functional.gelu,
+        attention_probs_dropout_prob: float = 0.0,
+        layer_norm_eps: float = 1e-12,
+        **kwargs: Any,
+    ):
+        super().__init__()
+        self.layer = nn.ModuleList(
+            [
+                TransformerLayer(
+                    hidden_size=hidden_size,
+                    num_attention_heads=num_attention_heads,
+                    hidden_dropout_prob=hidden_dropout_prob,
+                    intermediate_size=intermediate_size,
+                    intermediate_activation=intermediate_activation,
+                    attention_probs_dropout_prob=attention_probs_dropout_prob,
+                    layer_norm_eps=layer_norm_eps,
+                )
+                for _ in range(num_hidden_layers)
+            ]
+        )
+        self.hidden_size = hidden_size
+
+    def forward(
+        self,
+        hidden_states: Tensor,
+        attention_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+    ):
+        all_hidden_states = []
+        all_self_attentions = []
+
+        for i, layer_module in enumerate(self.layer):
+            all_hidden_states.append(hidden_states)
+
+            layer_head_mask = head_mask[i] if head_mask is not None else None
+            layer_outputs = layer_module(hidden_states, attention_mask, layer_head_mask)
+
+            hidden_states = layer_outputs[0]
+
+            all_self_attentions.append(layer_outputs[1])
+
+        all_hidden_states.append(hidden_states)
+
+        all_hidden_states = tuple(all_hidden_states)
+        all_self_attentions = tuple(all_self_attentions)
+
+        return TransformerOutput(
+            last_hidden_state=hidden_states,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attentions,
+        )
+
+
+class TransformerWithoutEmbeddings(nn.Module):
+    # TODO(asg): Add support for pretrained checkpoint loading
+    def __init__(
+        self,
+        encoder: nn.Module,
+        layernorm: nn.Module,
+        pooler: nn.Module,
+        weight_init_fn: Optional[Callable] = None,
+        initializer_range: float = 0.02,
+        use_cls_token: bool = True,
+        **kwargs: Any,
+    ):
+        super().__init__()
+        self.encoder = encoder
+        self.layernorm = layernorm
+        self.pooler = pooler
+        if use_cls_token:
+            self.cls_token = nn.Parameter(torch.zeros(1, 1, encoder.hidden_size))
```

Comment:
@ankitade mypy cannot find the right overload for this `torch.zeros` call, it thinks this call has signature `(int, int, Union[Tensor, Module])` but `hidden_size` is an int member of `TransformerEncoder`. This probably starting showing up once it was moved under `flava.py` because mypy ignores `flava.py`. Do we assume that the encoder here will always be an instance of `TransformerEncoder`? Or can we give it any encoder? `hidden_size` is not a member of `nn.Module` so it looks like we expect a custom encoder class, in which case maybe we should change the type annotation in the init function

Reply:
does doing something like this fix the issue ?
hidden_size: int = encoder.hidden_size

Reply:
nope, same issue

Reply:
Convert to int before passing by `int(encoder.hidden_size)`?

Reply:
unfortunately, still the same issue

Reply:
added `#type: ignore` for now